src/charm.py

```diff
@@ -0,0 +1,323 @@
+        admin_switch = "--admin" if admin == "yes" else "--no-admin"
+
+        cmd = ["register_new_matrix_user",
+               "-u", user, "-p", password,
+               admin_switch, "-c", self._SYNAPSE_CONFIG_PATH,
+               "http://localhost:8008"]
+        process = self._container().exec(
+            cmd,
+            working_dir="/data",
+            environment=self._populate_synapse_env_settings(),
+        )
+        try:
+            stdout, stderr = process.wait_output()
+            result = Result(
+                    return_code=0, stdout=stdout, stderr=stderr)
+        except ops.pebble.ExecError as error:
+            result = Result(error.exit_code, error.stdout, error.stderr)
+        return_code = result.return_code
+        logger.debug(
+                  "Run command: %s, return code %s\nstdout: %s\nstderr:%s",
+                  cmd,
+                  return_code,
+                  result.stdout,
+                  result.stderr,
+         )
+
+    def _run_generate_synapse(self):
+        """Runs the generate command for synapse"""
+        Result = collections.namedtuple(
+                "CommandExecResult", "return_code stdout stderr")
+
+        cmd = ["/start.py", "generate"]
+        process = self._container().exec(
+            cmd,
+            working_dir="/data",
+            environment=self._populate_synapse_env_settings(),
+        )
+        try:
+            stdout, stderr = process.wait_output()
+            result = Result(
+                    return_code=0, stdout=stdout, stderr=stderr)
+        except ops.pebble.ExecError as error:
+            result = Result(error.exit_code, error.stdout, error.stderr)
+        return_code = result.return_code
+        logger.debug(
+                  "Run command: %s, return code %s\nstdout: %s\nstderr:%s",
+                  cmd,
+                  return_code,
+                  result.stdout,
+                  result.stderr,
+         )
+
+    def _run_migrate_synapse(self):
+        """Runs the migrate command for synapse"""
+        Result = collections.namedtuple(
+                "CommandExecResult", "return_code stdout stderr")
+        cmd = ["/start.py", "migrate_config"]
+        process = self._container().exec(
+            cmd,
+            working_dir="/data",
+            environment=self._populate_synapse_env_settings(),
+        )
+        try:
+            stdout, stderr = process.wait_output()
+            result = Result(
+                    return_code=0, stdout=stdout, stderr=stderr)
+        except ops.pebble.ExecError as error:
+            result = Result(error.exit_code, error.stdout, error.stderr)
+        return_code = result.return_code
+        logger.debug(
+                  "Run command: %s, return code %s\nstdout: %s\nstderr:%s",
+                  cmd,
+                  return_code,
+                  result.stdout,
+                  result.stderr,
+         )
+
+    def _current_synapse_config(self):
```

Comment:
type hints? Or for the future?

Reply:
I'm tempted to say for the future, although it should be the very near future while the code base is still small. I expect a lot of things to change with the received feedback, so I suggest to pause for until the PoC is done.